File: testdata/fixtures/src/sample.php
<?php
/**
 * sample.php — comprehensive PHP syntax fixture for parser testing.
 * Covers: namespaces, interfaces, abstract classes, traits, enums, generics
 * via docblocks, first-class callables, fibers, match expressions, named args,
 * readonly properties, union/intersection types, nullsafe operator.
 */

declare(strict_types=1);

namespace Sample;

use DateTimeImmutable;
use Generator;
use InvalidArgumentException;
use RuntimeException;
use Throwable;

// -------------------------------------------------------------------------- //
// Constants
// -------------------------------------------------------------------------- //

const MAX_RETRIES = 3;
const DEFAULT_TIMEOUT = 30.0;

// -------------------------------------------------------------------------- //
// Enums (PHP 8.1+)
// -------------------------------------------------------------------------- //

enum Status: string
{
    case Pending = 'pending';
    case Running = 'running';
    case Done    = 'done';
    case Failed  = 'failed';

    public function isTerminal(): bool
    {
        return match ($this) {
            self::Done, self::Failed => true,
            default                 => false,
        };
    }

    public function label(): string
    {
        return match ($this) {
            self::Pending => 'Waiting to start',
            self::Running => 'In progress',
            self::Done    => 'Completed',
            self::Failed  => 'Failed',
        };
    }
}

// -------------------------------------------------------------------------- //
// Interfaces
// -------------------------------------------------------------------------- //

/** @template T */
interface Repository
{
    /** @return T|null */
    public function findById(string $id): mixed;

    /** @return list<T> */
    public function list(): array;

    /** @param T $entity @return T */
    public function save(mixed $entity): mixed;

    public function delete(string $id): bool;
}

interface Validatable
{
    /** @return list<string> */
    public function validationErrors(): array;

    public function isValid(): bool;
}

// -------------------------------------------------------------------------- //
// Traits
// -------------------------------------------------------------------------- //

trait HasTimestamps
{
    private DateTimeImmutable $createdAt;
    private DateTimeImmutable $updatedAt;

    public function initTimestamps(): void
    {
        $now = new DateTimeImmutable();
        $this->createdAt = $now;
        $this->updatedAt = $now;
    }

    public function getCreatedAt(): DateTimeImmutable
    {
        return $this->createdAt;
    }

    public function touch(): void
    {
        $this->updatedAt = new DateTimeImmutable();
    }
}

trait Serializable
{
    public function toArray(): array
    {
        return get_object_vars($this);
    }

    public function toJson(): string
    {
        return json_encode($this->toArray(), JSON_THROW_ON_ERROR);
    }
}

// -------------------------------------------------------------------------- //
// Abstract class
// -------------------------------------------------------------------------- //

abstract class BaseEntity
{
    use HasTimestamps;
    use Serializable;

    public function __construct(public readonly string $id)
    {
        $this->initTimestamps();
    }

    abstract public function validate(): void;
}

// -------------------------------------------------------------------------- //
// Readonly class (PHP 8.2+) / readonly properties
// -------------------------------------------------------------------------- //

class Point
{
    public function __construct(
        public readonly float $x,
        public readonly float $y,
    ) {}

    public static function origin(): self
    {
        return new self(0.0, 0.0);
    }

    public function distanceTo(self $other): float
    {
        return sqrt(($this->x - $other->x) ** 2 + ($this->y - $other->y) ** 2);
    }

    public function __toString(): string
    {
        return "({$this->x}, {$this->y})";
    }
}

// -------------------------------------------------------------------------- //
// Domain entity
// -------------------------------------------------------------------------- //

class User extends BaseEntity implements Validatable
{
    public function __construct(
        string $id,
        public string $name,
        public string $email,
        public string $role = 'member',
    ) {
        parent::__construct($id);
    }

    public function validate(): void
    {
        if (trim($this->name) === '') {
            throw new InvalidArgumentException("name must not be blank, got: {$this->name}");
        }
        if (!str_contains($this->email, '@')) {
            throw new InvalidArgumentException("email '{$this->email}' is invalid");
        }
    }

    public function validationErrors(): array
    {
        $errors = [];
        if (trim($this->name) === '') $errors[] = 'name is required';
        if (!str_contains($this->email, '@')) $errors[] = "email '{$this->email}' is invalid";
        return $errors;
    }

    public function isValid(): bool
    {
        return $this->validationErrors() === [];
    }
}

// -------------------------------------------------------------------------- //
// Generic-ish in-memory repository
// -------------------------------------------------------------------------- //

/** @template T of BaseEntity */
class MemoryRepository implements Repository
{
    /** @var array<string, T> */
    private array $store = [];

    public function findById(string $id): mixed
    {
        return $this->store[$id] ?? null;
    }

    public function list(): array
    {
        return array_values($this->store);
    }

    public function save(mixed $entity): mixed
    {
        $this->store[$entity->id] = $entity;
        return $entity;
    }

    public function delete(string $id): bool
    {
        if (!isset($this->store[$id])) return false;
        unset($this->store[$id]);
        return true;
    }

    /** @param callable(T): bool $predicate @return list<T> */
    public function findWhere(callable $predicate): array
    {
        return array_values(array_filter($this->store, $predicate));
    }
}

// -------------------------------------------------------------------------- //
// Exception hierarchy
// -------------------------------------------------------------------------- //

class AppException extends RuntimeException
{
    public function __construct(
        string $message,
        public readonly string $code,
        ?Throwable $previous = null,
    ) {
        parent::__construct($message, 0, $previous);
    }
}

class NotFoundException extends AppException
{
    public function __construct(string $resource, string $id)
    {
        parent::__construct("{$resource} with id='{$id}' not found", 'NOT_FOUND');
    }
}

class ValidationException extends AppException
{
    public function __construct(/** @param list<string> */ public readonly array $errors)
    {
        parent::__construct('Validation failed: ' . implode(', ', $errors), 'VALIDATION_ERROR');
    }
}

// -------------------------------------------------------------------------- //
// First-class callables & match
// -------------------------------------------------------------------------- //

function pipeline(mixed $value, callable ...$fns): mixed
{
    return array_reduce($fns, fn ($carry, $fn) => $fn($carry), $value);
}

function classify(mixed $value): string
{
    return match (true) {
        is_null($value)              => 'null',
        is_bool($value)              => 'boolean: ' . ($value ? 'true' : 'false'),
        is_int($value) && $value < 0 => 'negative int',
        is_int($value)               => 'non-negative int',
        is_string($value)            => 'string of length ' . strlen($value),
        is_array($value)             => 'array of ' . count($value),
        is_object($value)            => 'object ' . get_class($value),
        default                      => 'unknown',
    };
}

// -------------------------------------------------------------------------- //
// Generators
// -------------------------------------------------------------------------- //

function fibonacci(): Generator
{
    [$a, $b] = [0, 1];
    while (true) {
        yield $a;
        [$a, $b] = [$b, $a + $b];
    }
}

function take(int $n, Generator $gen): array
{
    $result = [];
    for ($i = 0; $i < $n; $i++) {
        $result[] = $gen->current();
        $gen->next();
    }
    return $result;
}

// -------------------------------------------------------------------------- //
// Fibers (PHP 8.1+)
// -------------------------------------------------------------------------- //

function makeAsyncTask(string $name): \Fiber
{
    return new \Fiber(function () use ($name): string {
        echo "Starting {$name}\n";
        $result = \Fiber::suspend("midpoint of {$name}");
        echo "Resumed {$name} with: {$result}\n";
        return "done: {$name}";
    });
}

// -------------------------------------------------------------------------- //
// Nullsafe operator & named arguments
// -------------------------------------------------------------------------- //

function formatUser(?User $user): string
{
    $name = $user?->name ?? 'Guest';
    $upper = mb_strtoupper(string: $name, encoding: 'UTF-8');
    return "Hello, {$upper}!";
}

// -------------------------------------------------------------------------- //
// Union & intersection types
// -------------------------------------------------------------------------- //

function acceptNumber(int|float $n): float
{
    return (float) $n;
}

interface Countable2 { public function count(): int; }
interface Iterable2  { public function toArray(): array; }

function processCollection(Countable2&Iterable2 $col): void
{
    echo $col->count(), "\n";
}

// -------------------------------------------------------------------------- //
// Entry point
// -------------------------------------------------------------------------- //

$user = new User(id: 'u1', name: 'Alice', email: 'alice@example.com');
$user->validate();

$repo = new MemoryRepository();
$repo->save($user);
echo $repo->findById('u1')?->name, "\n";

$fibs = take(10, fibonacci());
echo implode(', ', $fibs), "\n";

$p1 = Point::origin();
$p2 = new Point(3.0, 4.0);
echo "distance: ", $p2->distanceTo($p1), "\n";

$fiber = makeAsyncTask('task-1');
$mid = $fiber->start();
$fiber->resume('hello');

File: testdata/fixtures/src/sample.cs
/**
 * sample.cs — comprehensive C# syntax fixture for parser testing.
 * Covers: records, interfaces, generics, LINQ, async/await, pattern matching,
 * extension methods, delegates, events, indexers, operator overloading,
 * nullable reference types, primary constructors (C#12), required members.
 */

#nullable enable

using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;
using System.Runtime.CompilerServices;

namespace Sample;

// -------------------------------------------------------------------------- //
// Constants
// -------------------------------------------------------------------------- //

public static class Constants
{
    public const int MaxRetries = 3;
    public const string DefaultRole = "member";
}

// -------------------------------------------------------------------------- //
// Enums
// -------------------------------------------------------------------------- //

public enum Status
{
    Pending,
    Running,
    Done,
    Failed
}

public static class StatusExtensions
{
    public static bool IsTerminal(this Status s) =>
        s is Status.Done or Status.Failed;

    public static string Describe(this Status s) => s switch
    {
        Status.Pending => "Waiting to start",
        Status.Running => "In progress",
        Status.Done    => "Completed",
        Status.Failed  => "Failed",
        _              => throw new ArgumentOutOfRangeException(nameof(s), s, null)
    };
}

// -------------------------------------------------------------------------- //
// Records
// -------------------------------------------------------------------------- //

public record Point(double X, double Y)
{
    public double DistanceTo(Point other) =>
        Math.Sqrt(Math.Pow(X - other.X, 2) + Math.Pow(Y - other.Y, 2));

    public static Point Origin => new(0, 0);
}

public record struct Size(double Width, double Height)
{
    public double Area => Width * Height;
}

// -------------------------------------------------------------------------- //
// Interfaces
// -------------------------------------------------------------------------- //

public interface IRepository<T, TId>
{
    Task<T?> FindByIdAsync(TId id, CancellationToken ct = default);
    Task<IReadOnlyList<T>> ListAsync(CancellationToken ct = default);
    Task<T> SaveAsync(T entity, CancellationToken ct = default);
    Task<bool> DeleteAsync(TId id, CancellationToken ct = default);
}

public interface IValidator<T>
{
    bool Validate(T value, out IReadOnlyList<string> errors);
}

// -------------------------------------------------------------------------- //
// Abstract class
// -------------------------------------------------------------------------- //

public abstract class BaseEntity
{
    public required string Id { get; init; }
    public DateTime CreatedAt { get; } = DateTime.UtcNow;
    public DateTime UpdatedAt { get; private set; } = DateTime.UtcNow;

    protected void Touch() => UpdatedAt = DateTime.UtcNow;

    public override bool Equals(object? obj) =>
        obj is BaseEntity e && Id == e.Id;

    public override int GetHashCode() => Id.GetHashCode();
}

// -------------------------------------------------------------------------- //
// Domain entity
// -------------------------------------------------------------------------- //

public sealed class User : BaseEntity
{
    public required string Name { get; set; }
    public required string Email { get; set; }
    public string Role { get; set; } = Constants.DefaultRole;

    public void UpdateName(string name)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(name);
        Name = name;
        Touch();
    }
}

// -------------------------------------------------------------------------- //
// Generic in-memory repository
// -------------------------------------------------------------------------- //

public sealed class MemoryRepository<T> : IRepository<T, string>
    where T : BaseEntity
{
    private readonly Dictionary<string, T> _store = new();
    private readonly SemaphoreSlim _lock = new(1, 1);

    public async Task<T?> FindByIdAsync(string id, CancellationToken ct = default)
    {
        await _lock.WaitAsync(ct);
        try { return _store.TryGetValue(id, out var e) ? e : null; }
        finally { _lock.Release(); }
    }

    public async Task<IReadOnlyList<T>> ListAsync(CancellationToken ct = default)
    {
        await _lock.WaitAsync(ct);
        try { return _store.Values.ToList().AsReadOnly(); }
        finally { _lock.Release(); }
    }

    public async Task<T> SaveAsync(T entity, CancellationToken ct = default)
    {
        await _lock.WaitAsync(ct);
        try { _store[entity.Id] = entity; return entity; }
        finally { _lock.Release(); }
    }

    public async Task<bool> DeleteAsync(string id, CancellationToken ct = default)
    {
        await _lock.WaitAsync(ct);
        try { return _store.Remove(id); }
        finally { _lock.Release(); }
    }
}

// -------------------------------------------------------------------------- //
// LINQ & extension methods
// -------------------------------------------------------------------------- //

public static class EnumerableExtensions
{
    public static IEnumerable<T> WhereNotNull<T>(this IEnumerable<T?> source) where T : class =>
        source.Where(x => x is not null)!;

    public static IEnumerable<(T Item, int Index)> Indexed<T>(this IEnumerable<T> source) =>
        source.Select((item, i) => (item, i));

    public static IReadOnlyDictionary<TKey, IReadOnlyList<T>> GroupByReadOnly<T, TKey>(
        this IEnumerable<T> source, Func<T, TKey> keySelector) where TKey : notnull =>
        source.GroupBy(keySelector)
              .ToDictionary(g => g.Key, g => (IReadOnlyList<T>)g.ToList());
}

// -------------------------------------------------------------------------- //
// Delegates, events, lambdas
// -------------------------------------------------------------------------- //

public delegate TResult Transformer<in T, out TResult>(T input);

public sealed class EventBus
{
    private readonly Dictionary<Type, List<Delegate>> _handlers = new();

    public event EventHandler<Exception>? ErrorOccurred;

    public void Subscribe<T>(Action<T> handler)
    {
        var type = typeof(T);
        if (!_handlers.ContainsKey(type)) _handlers[type] = new();
        _handlers[type].Add(handler);
    }

    public void Publish<T>(T message)
    {
        if (!_handlers.TryGetValue(typeof(T), out var list)) return;
        foreach (var h in list)
        {
            try { ((Action<T>)h)(message); }
            catch (Exception ex) { ErrorOccurred?.Invoke(this, ex); }
        }
    }
}

// -------------------------------------------------------------------------- //
// Async streams (IAsyncEnumerable)
// -------------------------------------------------------------------------- //

public static class AsyncStreams
{
    public static async IAsyncEnumerable<int> FibonacciAsync(
        int count,
        [EnumeratorCancellation] CancellationToken ct = default)
    {
        int a = 0, b = 1;
        for (int i = 0; i < count; i++)
        {
            ct.ThrowIfCancellationRequested();
            yield return a;
            await Task.Yield();
            (a, b) = (b, a + b);
        }
    }
}

// -------------------------------------------------------------------------- //
// Pattern matching
// -------------------------------------------------------------------------- //

public static class Classifier
{
    public static string Classify(object? value) => value switch
    {
        null                   => "null",
        true                   => "boolean true",
        false                  => "boolean false",
        int n when n < 0       => $"negative int {n}",
        int n                  => $"non-negative int {n}",
        string { Length: 0 }   => "empty string",
        string s               => $"string of length {s.Length}",
        Point { X: 0, Y: 0 }  => "origin point",
        Point p                => $"point at ({p.X},{p.Y})",
        IEnumerable<int> list  => $"int collection with {list.Count()} items",
        _                      => $"unknown: {value.GetType().Name}"
    };
}

// -------------------------------------------------------------------------- //
// Operator overloading
// -------------------------------------------------------------------------- //

public readonly struct Money
{
    public decimal Amount { get; }
    public string Currency { get; }

    public Money(decimal amount, string currency)
    {
        if (amount < 0) throw new ArgumentOutOfRangeException(nameof(amount), amount, "must be non-negative");
        Amount = amount;
        Currency = currency;
    }

    public static Money operator +(Money a, Money b)
    {
        if (a.Currency != b.Currency)
            throw new InvalidOperationException($"Currency mismatch: {a.Currency} vs {b.Currency}");
        return new(a.Amount + b.Amount, a.Currency);
    }

    public static bool operator >(Money a, Money b)  => a.Amount > b.Amount;
    public static bool operator <(Money a, Money b)  => a.Amount < b.Amount;
    public static bool operator >=(Money a, Money b) => a.Amount >= b.Amount;
    public static bool operator <=(Money a, Money b) => a.Amount <= b.Amount;

    public override string ToString() => $"{Amount:F2} {Currency}";
}

// -------------------------------------------------------------------------- //
// Indexers
// -------------------------------------------------------------------------- //

public sealed class Matrix
{
    private readonly double[,] _data;

    public Matrix(int rows, int cols) => _data = new double[rows, cols];

    public double this[int row, int col]
    {
        get => _data[row, col];
        set => _data[row, col] = value;
    }

    public int Rows => _data.GetLength(0);
    public int Cols => _data.GetLength(1);
}

// -------------------------------------------------------------------------- //
// Error types
// -------------------------------------------------------------------------- //

public sealed class AppException : Exception
{
    public string Code { get; }

    public AppException(string message, string code, Exception? inner = null)
        : base(message, inner)
    {
        Code = code;
    }
}

// -------------------------------------------------------------------------- //
// Entry point
// -------------------------------------------------------------------------- //

public static class Program
{
    public static async Task Main()
    {
        var user = new User { Id = Guid.NewGuid().ToString(), Name = "Alice", Email = "alice@example.com" };
        Console.WriteLine(user.Name);

        var status = Status.Done;
        Console.WriteLine(status.Describe());

        await foreach (var n in AsyncStreams.FibonacciAsync(10))
        {
            Console.Write($"{n} ");
        }
        Console.WriteLine();

        var p = new Point(3, 4);
        Console.WriteLine($"Distance from origin: {p.DistanceTo(Point.Origin)}");

        var money = new Money(10.50m, "USD") + new Money(5.25m, "USD");
        Console.WriteLine(money);
    }
}
